### PR TITLE
IGNITE-18952 Add the filter as a configuration parameter of a Distribution zone

### PR DESCRIPTION
**Pull request by alievmirza**
https://issues.apache.org/jira/browse/IGNITE-18952

### CODE REVIEW

File: modules/distribution-zones/build.gradle [hunk 1]
```diff
@@ -37,6 +37,11 @@ dependencies {
     implementation libs.jetbrains.annotations
     implementation libs.fastutil.core
     implementation libs.auto.service.annotations
+    implementation(libs.jsonpath.core) {
+        //IDEA test runner doesn't apply Gradle dependency resolve strategy, this is just not implemented
```
Comment (kgusakov) [inline modules/distribution-zones/build.gradle:41]: Not sure, that I understand how the first line connected with the second, could you explain?

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/configuration/DistributionZoneConfigurationSchema.java [hunk 2]
```diff
@@ -67,4 +67,8 @@ public class DistributionZoneConfigurationSchema {
     @Range(min = 0)
     @Value(hasDefault = true)
     public int dataNodesAutoAdjustScaleDown = INFINITE_TIMER_VALUE;
+
+    @ValidFilter
+    @Value(hasDefault = true)
+    public String filter = "$..*";
```
Comment (sanpwc) [inline modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/configuration/DistributionZoneConfigurationSchema.java:73]: What "$..*" means? Any String? Let's add a comment with an explanation.

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZonesUtil.java [hunk 3]
```diff
@@ -347,4 +349,20 @@ public static DistributionZoneConfiguration getZoneById(DistributionZonesConfigu
...
+    /**
+     * Check if a passed filter is a valid {@link JsonPath} query.
+     *
+     * @param filter Filter.
+     * @return {@code true} if the passed filter is a valid filter, {@code false} otherwise.
+     */
+    public static boolean validate(String filter) {
+        try {
+            JsonPath.compile(filter);
+        } catch (InvalidPathException ignored) {
```
Comment (sanpwc) [inline modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZonesUtil.java:362]: What about validation details? It's useful to propagate validation details to the user.

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/configuration/DistributionZoneConfigurationSchema.java (hunk 2 above)
Comment (sanpwc) [inline modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/configuration/DistributionZoneConfigurationSchema.java:73]: And what if user will try to change the filter from some to "anything"? Do you expect them to specify "$..*"?

File: modules/distribution-zones/build.gradle (hunk 1 above)
Comment (alievmirza) [reply modules/distribution-zones/build.gradle:41]: This is a copy-paste from the other module where JsonPath was introduced, I think it's important to exclude this libs here as well

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/configuration/DistributionZoneConfigurationSchema.java (hunk 2 above)
Comment (alievmirza) [reply modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/configuration/DistributionZoneConfigurationSchema.java:73]: This is a  com.jayway.jsonpath.JsonPath expression for including all attributes of nodes.

File: modules/distribution-zones/src/test/java/org/apache/ignite/internal/distributionzones/DistributionZoneManagerTest.java [hunk 4]
```diff
@@ -755,17 +740,118 @@ public void testTryDropDefaultZone() {
...
+        String notValidFilter = "['nodeAttributes[?(@.['region'] == 'EU')]";
+
+        assertThrowsWithCause(
+                () -> distributionZoneManager.alterZone(
+                        ZONE_NAME,
+                        new DistributionZoneConfigurationParameters.Builder(ZONE_NAME)
+                                .filter(notValidFilter).build()
+                        )
+                        .get(5, TimeUnit.SECONDS),
+                ConfigurationValidationException.class
```
[the remark this answers is not in the document]
Comment (alievmirza) [reply modules/distribution-zones/src/test/java/org/apache/ignite/internal/distributionzones/DistributionZoneManagerTest.java]: done

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZonesUtil.java (hunk 3 above)
Comment (alievmirza) [reply modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZonesUtil.java]: added

**Review by sanpwc — APPROVED** [review]

File: modules/distribution-zones/build.gradle (hunk 1 above)
Comment (kgusakov) [reply modules/distribution-zones/build.gradle:41]: I see, thanks!

**Pull request merged**